lib/elixir/src/elixir_with.erl

```diff
@@ -18,19 +18,27 @@ expand(Meta, Args, E) ->
         {Args, []}
     end,
 
-  {Expr, Opts} =
+  {DoExpr, BlockWithoutDo} =
     case lists:keytake(do, 1, Block) of
-      {value, {do, Do}, Rest} ->
-        {Do, Rest};
+      {value, {do, Do}, ElseRest} ->
+        {Do, ElseRest};
       false ->
         elixir_errors:compile_error(Meta, ?m(E, file),
           "missing do keyword in with")
     end,
+  {ElseExpr, Opts} =
+    case lists:keytake(else, 1, BlockWithoutDo) of
+      {value, {else, Else}, Rest} ->
+        {Else, Rest};
+      false ->
+        {nil, BlockWithoutDo}
+    end,
 
   {EOpts, EO} = elixir_exp:expand(Opts, E),
   {ECases, EC} = lists:mapfoldl(fun expand/2, EO, Cases),
-  {EExpr, _} = elixir_exp:expand(Expr, EC),
-  {{with, Meta, ECases ++ [[{do, EExpr} | EOpts]]}, E}.
+  {EExpr, _} = elixir_exp:expand(DoExpr, EC),
+  {EEExpr, _} = expand_elses(ElseExpr, E),
+  {{with, Meta, ECases ++ [[{do, EExpr} | EEExpr] ++ EOpts]}, E}.
```

Comment:
I do not think, that we do not need wrapping list `[...]`. Because, there is a natural reason for do and else to be in own wrapping list, because:

```elixir
with {:ok, width} <- Map.fetch(opts, :width), 
     {:ok, height} <- Map.fetch(opts, :height), do: {:ok, width * height}
# The same as:
with {:ok, width} <- Map.fetch(opts, :width), 
     {:ok, height} <- Map.fetch(opts, :height), [do: {:ok, width * height}]
# And the same as:
with {:ok, width} <- Map.fetch(opts, :width), 
     {:ok, height} <- Map.fetch(opts, :height), [{:do, {:ok, width * height}}]
```

And `do` and `else` are in own wrapping list and do not mixed with cases.